### ISSUE TITLE
Remote server not playing well with multiple clients/archives dir (permission issue since same files are used)

### ISSUE THREAD

**Issue by tlaurion** [tasket/wyng-backup#114]
Use case linked to PoC #100 

```
[user@dom0 ~]$ sudo wyng --meta-dir=/var/lib/wyng.backup-insurgo-RW --from=qubes-ssh://wyng-Insurgo-RW:rwuser@198.15.126.235/ --subdir=home/rwuser/q41_insurgo arch-init
Wyng 0.3.2 release 20220818

Metadata will be read from the archive;
Please check that the archive is in a trusted, secure condition!
Proceed? [y/N] y

```
No problem here. Doing my thing. Adding stuff, sending with tags and dedup etc.

Then when time comes to download changes from RO account:
```
[user@dom0 ~]$ sudo wyng --meta-dir=/var/lib/wyng.backup-insurgo --from=qubes-ssh://wyng-Insurgo:q41_insurgo@198.15.126.235/ --subdir=home/rwuser/q41_insurgo arch-init
Wyng 0.3.2 release 20220818

Metadata will be read from the archive;
Please check that the archive is in a trusted, secure condition!
Proceed? [y/N] y
Traceback (most recent call last):
  File "/usr/bin/wyng", line 2918, in <module>
    aset         = get_configs_remote()
  File "/usr/bin/wyng", line 773, in get_configs_remote
    do_exec([dest_run_args(desttype, [destcd + bkdir
  File "/usr/bin/wyng", line 1154, in do_exec
    raise subprocess.CalledProcessError(err.returncode, err.args)
subprocess.CalledProcessError: Command '['/bin/tar', '-xzf', '-']' returned non-zero exit status 2.
```

```
[user@dom0 ~]$ cat /tmp/wyng/err.log 
--+--
--+--
--+--
rm: cannot remove '/tmp/wyng-rpc/dest_helper.py': Permission denied
rm: cannot remove '/tmp/wyng-rpc/tar.lst': Permission denied
--+--
--+--
Traceback (most recent call last):
  File "/tmp/wyng-rpc/dest_helper.py", line 150, in <module>
    with open(tmpdir+"/tar.lst","w") as tarlstf:
PermissionError: [Errno 13] Permission denied: '/tmp/wyng-rpc/tar.lst'

gzip: stdin: unexpected end of file
/bin/tar: Child returned status 1
/bin/tar: Error is not recoverable: exiting now
--+--
```

Workaround: 
- Go on server with root account and delete /tmp/wyng-rpc directory.

Needed implementation:
- Probably create /tmp/name_of_wyng-metadir or something so that those are not conflicting
- Above solution might not be sufficient if multiple users are attempting to use the server at the same time. Maybe create /tmp/name_of_dir with mktemp?

This might be blocker in PoC.

**Comment by tasket** [comment]
Already looking at making tmp dirs unique.  Fix coming soon.

**Comment by tasket** [comment]
Related issues #15, #45

**Comment by tasket** [comment]
@tlaurion There is now a workaround to allow multiple remote helpers in the 'fix03' branch.  It is based on simple (large) random number assignment, so there is a tiny chance of failure.  The alternative would be to have the remote assign a dir name and then store that in a local variable, but I'm not comfortable using an untrusted filesystem parsed string at this time.

**Comment by tasket** [comment]
I went ahead and used a solution for sanitizing the remote tmp dir name... for complete collision avoidance.

**Comment by t-gh-ctrl** [comment]
FWIW I ran into that issue a couple times while doing tests with beta 20250107: I eventually had to `rm -fr /tmp/wyngrpc` on the remote host otherwise I'd get a "cleanup() ..." error after sending volumes.

steps to reproduce:

- on remote host: `rm -fr /tmp/wyngrpc`

- configure wyng properly and send a volume to `qubes-ssh://<vm>:alice@remotehost/alice-bkp`

- on remotehost, `/tmp/wyngrpc` now exists and is owned by alice

- change the destination to `qubes-ssh://<vm>:bob@remotehost/alice-bkp` and send a volume; wyng will (rightfully) complain that the archive can't be written to (since bob tries to access alice's files) and will exit:

```text
touch: cannot touch 'archive.dat': Permission denied
[0, 1]

Destination not ready to receive commands.
```

- on remotehost, `/tmp/wyngrpc` is still owned by alice, and a "dangling" `/tmp/wyngrpc/tmp<random>` directory is owned by bob.

- switch the destination back to `qubes-ssh://<vm>:alice@remotehost/alice-bkp`, and send a volume. The volume seems to be properly sent but there's a cleanup() error:

```text
Wyng 0.8 beta release 20250107
Encrypted archive 'qubes-ssh://vm:alice@remotehost:alice-bkp'
Last updated 2025-01-17 11:03:05.642150 (+02:00)

Preparing snapshots in '/dev/qubes_dom0/'...
Acquiring deltas.

Sending backup session 20250117-123231:
——————————————————————————————————————————
  no change |  -  | vm-blah-private
——————————————————————————————————————————
1 volumes, 5724——>0 MB in 0.8 seconds.
Cleanup error: CalledProcessError(1, [['/usr/bin/qvm-run', '--no-color-stderr', '--no-color-output', '--nogui', '--no-filter-escape-chars', '-p', 'infra', '/usr/bin/ssh -x -T -e none -o ControlPath=~/.ssh/ctrl-%C -o ControlMaster=auto -o ControlPersist=300 -o ServerAliveInterval=120 -o ConnectTimeout=60 -o Compression=no ssh://z690@fw.leshten.priv: "$(/usr/bin/cat /tmp/wyngrpc/tmp8foy02_q)"']])
```

- on remote host, `rm -fr /tmp/wyngrpc`; sending is now OK.

**Comment by tasket** [comment]
Related to #231 since they are rooted in the same part of _class Destination:_ code.

**Comment by t-gh-ctrl** [comment]
I got that issue again yesterday while doing tests on other issues and had to `rm -fr /tmp/wyngrpc` on the ssh host to fix it (I was in a hurry - I should have looked at what was happening instead). I'll post back if it happens again.

IIRC I interrupted wyng a couple times (ctrl-c) but I don't remember on which host -  the host that was backing up vms, or the host that was restoring them, or maybe both.

FWIW, I'm using almost identical destinations on both the master and slave hosts, only the ssh user changes, so that the "slave" host (/laptop) can't mess up with the "master" host's archive on the ssh host.

Ie.
```
master: dest=qubes-ssh://vm-backup/masteruser@sshhost/sshdir
slave:  dest=qubes-ssh://vm-backup/slaveuser@sshhost/sshdir
```

**Comment by tasket** [comment]
Two questions:

1. Is the error still occurring?
2. If so, are both "alice" and "bob" going through the same "vm-backup" VM (i.e. this is not two different vm-backup VMs on two different machines)?

**Comment by t-gh-ctrl** [comment]
> 1. Is the error still occurring

No, but I haven't done extensive tests with wyng like I did last week. I'll backup my archive next week and do tests of things that can go wrong like interrupting wyng on the hosts (which is what seems to have caused the error last time), restarting the remote host while wyng is backuping or restoring, etc.

Feel free to close the issue and I'll post a comment if the error re-occurs.

(2- "alice" and "bob" don't go through the same vm-backup ; they have different names, and are on different hosts).
